presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java

```diff
@@ -0,0 +1,449 @@
+
+import com.facebook.presto.spi.type.NamedTypeSignature;
+import com.facebook.presto.spi.type.ParameterKind;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.facebook.presto.type.UnknownType;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.type.TypeCalculation.calculateLiteralValue;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getLast;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+public class DefaultSignatureBinder
+        extends SignatureBinder
+{
+    protected final Map<String, TypeVariableConstraint> typeVariableConstraints;
+
+    public DefaultSignatureBinder(TypeManager typeManager, Signature declaredSignature, boolean allowCoercion)
+    {
+        super(typeManager, declaredSignature, allowCoercion);
+        this.typeVariableConstraints = declaredSignature.getTypeVariableConstraints()
+                .stream()
+                .collect(toMap(TypeVariableConstraint::getName, t -> t));
+    }
+
+    @Override
+    public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+    {
+        Optional<BoundVariables> boundVariables = bindVariables(actualArgumentTypes);
+        if (!boundVariables.isPresent()) {
+            return Optional.empty();
+        }
+        return Optional.of(bindVariables(declaredSignature, boundVariables.get(), actualArgumentTypes.size()));
+    }
+
+    @Override
+    public Optional<BoundVariables> bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+    {
+        BoundVariables.Builder variableBinder = BoundVariables.builder();
+
+        TypeSignature expectedReturnTypeSignature = declaredSignature.getReturnType();
+        if (!bind(expectedReturnTypeSignature, actualReturnType, variableBinder)) {
+            return Optional.empty();
+        }
+
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, variableBinder)) {
+            return Optional.empty();
+        }
+
+        calculateVariableValuesForLongConstraints(variableBinder);
+        BoundVariables boundVariables = variableBinder.build();
+        if (!allTypeVariablesBound(boundVariables)) {
+            return Optional.empty();
+        }
+        return Optional.of(boundVariables);
+    }
+
+    public static Signature bindVariables(Signature signature, BoundVariables boundVariables, int arity)
```

Comment:
It's strange to have both static and non-static methods (also with different return value) with the same name in the same class. It suggest that those methods do the same logically.

I would suggest to refactor `bindVariables` that do not take `actualReturnType` into:
`Optional<BoundVariables> bindArgumentVariables(...)` or similar. You could then use those methods in `bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)`.

Static methods should go to the top of class.